### PR TITLE
Feature/m0127 cc 568 tutorial build micro ej firmware

### PR DESCRIPTION
**Pull request by gjadi-mej**

### CODE REVIEW

**Comment by gjadi-mej** [comment]
@thomasponsmicroej You're right about the diagram and "Library repository", unfortunately I was able to find the original file for this image.

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+2. Build the MicroEJ Application into a ``microejapp.o`` using MicroEJ SDK
+
+3. Compile the BSP C sources into ``.o`` using the C toolchain
+
+4. the BSP (``.o``) and the MicroEJ Application (``microejapp.o``) and
+   the MicroEJ Platform (``microejruntime.a``) are linked by the C toolchain to produce a
+   final ELF or binary called MicroEJ Firmware (e.g. ``application.out``).
+
+.. image:: images/build_microej_mono_sandbox_firmware_numbered.PNG
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:190]: Use schema from https://documentation.cross/share/page/site/m0127customercare/document-details?nodeRef=workspace://SpacesStore/05857911-5344-4c8b-9d02-3b2a34562975

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+* Rebuild the MicroEJ Platform in the following cases:
+
+  * When the MicroEJ Architecture (``.xpf``) changes.
+
+  * When a MicroEJ Foundation Library (``.xpfp``) changes.
+
+  * When a Foundation Library changes, either when
+
+    * The public API (``.java`` or ``.h``) change.
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:207]:
```suggestion
    * The public API (``.java`` or ``.h``) changes.
```

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+A good knowledge of the :ref:`chapter-glossary` is required.
+
+The Components
+--------------
+
+As depicted in the following image, several resources and tools are
+used to build a MicroEJ Firmware.
+
+.. image:: images/qa_resources-v3.png
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:19]: "MEJ32 Architecture" -> "MicroEJ Architecture"
"MEJ32 Repository" -> "MicroEJ Repository"
"MEJ32 Arch/Compiler" -> "MicroEJ Architecture"

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+===============================================================
+
+A MicroEJ Firmware is built from several input resources and tools.
+Each component has a specific version that must be carefully respected
+in order to build a firmware.
+
+This document describes the components and the process involved in the
+build of a MicroEJ Firmware.
+
+A good knowledge of the :ref:`chapter-glossary` is required.
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:11]:
```suggestion
Good knowledge of the :ref:`chapter-glossary` is required.
```

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+	:align: center
+
+MicroEJ Architecture (.xpf, .xpfp)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A MicroEJ Architecture contains the runtime port to a target
+instruction set (ISA) and a C compiler (CC) and MicroEJ Foundation
+Libraries.
+
+The MicroEJ Architectures are distributed into 2 formats:
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:29]:
```suggestion
The MicroEJ Architectures are distributed into two formats:
```

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+* EVAL: evaluation license with runtime limitations (explained in SDK
+  developer guide).
+* DEV: production license (only MicroEJ sales & Customer Care team distribute this version).
+
+The list of supported MicroEJ Architectures can be found here
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:35]:
```suggestion
The supported MicroEJ Architectures are listed here
```

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+* ``<platform>-configuration``: The configuration of the MicroEJ
+  Platform
+* ``<platform>-bsp``: The C code for the board-specific files
+  (drivers).
+* ``<platform>-fp``: Front panel mockup for the simulator.
+
+See :ref:`new_platform_creation` to learn how to create a MicroEJ
+Platform using a MicroEJ Platform Source project.
+
+Depending on the requirements of the project, the MicroEJ Platform can
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:71]:
```suggestion
Depending on the project's requirements, the MicroEJ Platform can
```

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+be connected in various ways to the BSP, see :ref:`bsp_connection` for
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:72]:
```suggestion
be connected in various ways to the BSP; see :ref:`bsp_connection` for
```

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+1. to either run on:
+
+  * a simulator (computer desktop),
+  * a device (actual embedded hardware).
+
+2. to setup:
+
+  * memory (example: Java heap, Java stack),
+  * foundation libraries,
+  * etc…
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:90]:
```suggestion
  * etc.
```

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+
+To run on a device, the application is compiled and optimized for a
+specific MicroEJ Platform.  It generates a ``microejapp.o`` (native
+object code) that is linked with the
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:94]:
```suggestion
object code) linked with the
```

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+``<platform>-bsp`` project.
+
+To import an existing MicroEJ Application as zip project in the SDK:
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:97]:
```suggestion
To import an existing MicroEJ Application as a zipped project in the SDK:
```

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+
+* Go to :guilabel:`File` > :guilabel:`Import…` > :guilabel:`General` >
+  :guilabel:`Existing Projects into Workspace` > :guilabel:`Select
+  archive file` > :guilabel:`Browse…`.
+* Select the zip of the project (e.g. ``x.zip``).
+* And select :guilabel:`Finish` import.
+
+See :ref:`microej.sample.applications` for more information on how to
+create, configure and develop a MicroEJ Application.
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:106]:
```suggestion
create, configure, and develop a MicroEJ Application.
```

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+  * the BSP C files (drivers).
+
+Module (Ivy) repository
+~~~~~~~~~~~~~~~~~~~~~~~
+
+A module (Ivy) repository provides the modules required to build
+MicroEJ Platforms and MicroEJ Applications.
+
+* The MicroEJ Central Repository is an online repository of software
+  artifacts (libraries, tools, …), see
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:125]:
```suggestion
  artifacts (libraries, tools, etc.), see
```

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+  https://repository.microej.com/. This repository can also be used as an offline repository, see https://developer.microej.com/central-repository/.
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:126]:
```suggestion
  https://repository.microej.com/. One can also use this repository as an offline repository, see https://developer.microej.com/central-repository/.
```

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+
+* (Optional) It can be extended with an offline repository (``.zip``)
+  that can be imported in the workspace (see
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:129]:
```suggestion
  and imported in the workspace (see
```

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+How to build
+------------
+
+The process to build a MicroEJ Firmware is two-fold:
+
+1. Build a MicroEJ Platform
+2. Build a MicroEJ Application into a MicroEJ Firmware
+
+The MicroEJ Application is compiled against the MicroEJ Platform to
+produce the MicroEJ Firmware that will be deployed on the target
```
Comment (jerome-leroux) [inline Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:167]:
```suggestion
produce the MicroEJ Firmware deployed on the target
```

**Review by jerome-leroux — CHANGES_REQUESTED** [review]

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst (hunk 15 above)
Comment (gjadi-mej) [reply Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:129]: Disagree. It's the offline repository that can be imported. Not the Central Repository to which the "It" refers to.

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst (hunk 14 above)
Comment (gjadi-mej) [reply Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:126]: Not in favor of this. We use the passive voice throughout the document, not the active one.

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst (hunk 3 above)
Comment (gjadi-mej) [reply Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:19]: Do you know where is located the source of this image? I couldn't find it.

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst [hunk 17]
```diff
@@ -0,0 +1,255 @@
+Understand how to build a MicroEJ Firmware and its dependencies
+===============================================================
+
+A MicroEJ Firmware is built from several input resources and tools.
+Each component has a specific version that must be carefully respected
+in order to build a firmware.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:6]: It emphasizes that one has to respect versions and dependencies between components. You can't make a firmware from incompatible components. I'll try to rephrase it.

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst (hunk 1 above)
Comment (gjadi-mej) [reply Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:200]: Done. Thanks.

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst (hunk 3 above)
Comment (gjadi-mej) [reply Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:19]: I redid it. WDYT?

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst (hunk 17 above)
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst]: Done. Thoughts on the rephrasing?

**Comment by gjadi-mej** [comment]
@thomasponsmicroej I redraw the diagram to integrate your suggestion. WDYT?

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst (hunk 17 above)
[the remark this answers is not in the document]
Comment (jerome-leroux) [reply Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst]: Good

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst (hunk 15 above)
Comment (jerome-leroux) [reply Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:129]: Yes

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst (hunk 14 above)
Comment (jerome-leroux) [reply Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:126]: OK. I tried to remove as much as possible the passive voice from the document.
The active voice can provide more clarity and brevity. Grammarly advises against using the passive voice.

File: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst (hunk 3 above)
Comment (jerome-leroux) [reply Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst:19]: Good

**Review by jerome-leroux — APPROVED** [review]

**Comment by thomasponsmicroej** [comment]
> 
> 
> @thomasponsmicroej I redraw the diagram to integrate your suggestion. WDYT?

ok for me

**Comment by thomasponsmicroej** [comment]
to continue with repositories naming, why mentionning ivy?  example: "Ivy Repository" in the last section or "Module (Ivy) Repository" in the repository section. I would use "Module repository" everywhere for simplicity and conformance with the glossary. I think we prefer to expose MMM rather than ivy or easyant when possible.

**Comment by gjadi-mej** [comment]
@thomasponsmicroej I replaced Module Repository + Ivy with MMM in the latest version of the diagram.

**Pull request merged**